src/compute-types/src/dataflows.rs

```diff
@@ -108,6 +112,38 @@ impl<T> DataflowDescription<Plan<T>, (), mz_repr::Timestamp> {
         as_of.try_step_forward().as_ref() == until.as_option()
     }
 
+    /// Returns the dataflow expiration, i.e, the timestamp beyond which diffs can be
+    /// dropped.
+    ///
+    /// Returns an empty timestamp if `replica_expiration` is unset or matches conditions under
+    /// which dataflow expiration should be disabled.
+    pub fn expire_dataflow_at(
+        &self,
+        replica_expiration: &Antichain<mz_repr::Timestamp>,
+    ) -> Antichain<mz_repr::Timestamp> {
+        let dataflow_expiration_desc = &self.dataflow_expiration_desc;
+
+        // Disable dataflow expiration if `replica_expiration` is unset, the current dataflow has a
+        // refresh schedule, has a transitive dependency with a refresh schedule, or the dataflow's
+        // timeline is not `Timeline::EpochMilliSeconds`.
+        if replica_expiration.is_empty()
+            || self.refresh_schedule.is_some()
+            || dataflow_expiration_desc.has_transitive_refresh_schedule
+            || !dataflow_expiration_desc.is_timeline_epoch_ms
+        {
+            return Antichain::default();
+        }
+
+        if let Some(upper) = &dataflow_expiration_desc.transitive_upper {
+            // Returns empty if `upper` is empty, else the max of `upper` and `replica_expiration`.
+            upper.join(replica_expiration)
+        } else {
+            replica_expiration.clone()
+        }
```

Comment:
We don't have to check for `replica_expiration` being empty, because that'll be handled correctly in the latter parts of the function.

```suggestion
        // Disable dataflow expiration if the current dataflow has a
        // refresh schedule, has a transitive dependency with a refresh schedule, or the dataflow's
        // timeline is not `Timeline::EpochMilliSeconds`.
        if self.refresh_schedule.is_some()
            || dataflow_expiration_desc.has_transitive_refresh_schedule
            || !dataflow_expiration_desc.is_timeline_epoch_ms
        {
            return Antichain::default();
        }

        if let Some(upper) = &dataflow_expiration_desc.transitive_upper {
            // Returns empty if `upper` is empty, else the max of `upper` and `replica_expiration`.
            upper.join(replica_expiration)
        } else {
            replica_expiration.clone()
        }
```

Reply:
I suggest keeping the extra check around. I added that to make it explicit that `dataflow_expiration` is enabled only if `replica_expiration` is enabled. Encoding this logic in an empty Antichain is less explicit for someone going through this code I think.

Plus we skip the upper comparison logic.